runtime/near-vm-runner/src/imports.rs

```diff
@@ -24,93 +25,105 @@ macro_rules! rust2wasm {
+                        let data = CALLER_CONTEXT.with(|caller_context| {
+                            unsafe {
+                                *caller_context.get()
+                            }
+                        });
+                        let logic: &mut VMLogic<'_> = unsafe { &mut *(data as *mut VMLogic<'_>) };
+                        match logic.$func( $( $arg_name as $arg_type, )* ) {
+                            Ok(result) => Ok(result as ($( rust2wasm!($returns) ),* ) ),
+                            Err(err) => {
+                                // Wasmtime doesn't have proper mechanism for wrapping custom errors
+                                // into traps. So, just store error into TLS and use special exit code here.
+                                EMBEDDER_ERROR.with(|embedder_error| {
+                                    *embedder_error.borrow_mut() = Some(err)
+                                });
+                                Err(Trap::i32_exit(239))
+                            }
                         }
                     }
-                }
-            )*
+                )*
             }
 
-            pub(crate) fn build_wasmer(memory: wasmer_runtime::memory::Memory, logic: &mut VMLogic<'_>) ->
-                wasmer_runtime::ImportObject {
+            #[allow(unused_variables)]
+            pub(crate) fn build_wasmer(
+                memory: wasmer_runtime::memory::Memory,
+                logic: &mut VMLogic<'_>,
+                protocol_version: ProtocolVersion,
+            ) -> wasmer_runtime::ImportObject {
                 let raw_ptr = logic as *mut _ as *mut c_void;
                 let import_reference = ImportReference(raw_ptr);
-                wasmer_runtime::imports! {
-                    move || {
-                        let dtor = (|_: *mut c_void| {}) as fn(*mut c_void);
-                        (import_reference.0, dtor)
-                    },
-                    "env" => {
-                        "memory" => memory,
-                        $(
-                            stringify!($func) => wasmer_runtime::func!(wasmer_ext::$func),
-                        )*
-                    },
-                }
+                let mut import_object = wasmer_runtime::ImportObject::new_with_data(move || {
+                    let dtor = (|_: *mut c_void| {}) as fn(*mut c_void);
+                    (import_reference.0, dtor)
+                });
+
+                let mut ns = wasmer_runtime_core::import::Namespace::new();
+                ns.insert("memory", memory);
+                $({
+                    $(#[cfg(feature = $feature_name)])*
+                    if true $(&& near_primitives::checked_feature!($feature_name, $feature, protocol_version))* {
+                        ns.insert(stringify!($func), wasmer_runtime::func!(wasmer_ext::$func));
+                    }
+                })*
+
+                import_object.register("env", ns);
```

Comment:
Why using a namespace object now?

Reply:
This is based on https://github.com/near/nearcore/pull/2842

Reply:
We need to do a dynamic construction of imports, because some feature might be on, but not enabled at the protocol version yet.

Reply:
Got it